File: AOC-2022/Day6.hs
module Main where
import Data.List
windows :: Int -> [a] -> [[a]]
windows _ [] = []
windows n xs | length xs < n = []
windows n xs = take n xs : windows n (tail xs)

main :: IO ()
main = do 
  print . (+4) . length . takeWhile ((/=) <*> nub) . windows 4 =<< readFile "input.txt"
  print . (+14) . length . takeWhile ((/=) <*> nub) . windows 14 =<< readFile "input.txt"

File: AOC-2022/Day6.rs
fn has_dups(xs: &[u8]) -> bool {
    let mut uniq = std::collections::HashSet::new();
    !xs.iter().all(move |x| uniq.insert(x))
}
fn main() {
    println!("{} {}", 
        std::fs::read("input.txt").unwrap().windows(4).take_while(|x|has_dups(x)).count() + 4
    ,   std::fs::read("input.txt").unwrap().windows(14).take_while(|x|has_dups(x)).count() + 14
    );
}
